### PR TITLE
Remove rental crate and implement the required functionality

### LINKED ISSUE #6312: Replace rental crate in sp_tracing
ref: PR #5826

Crate `rental` has just been announced as unmaintained. Needs to be removed and possibly use a different mechanism for capturing wasm traces. 

for more info, see: 

https://github.com/paritytech/substrate/pull/5826#issuecomment-638848885

### PR DESCRIPTION
**Pull request by mattrutherford**
Closes #6312

Introduces new struct `ProxiedSpan` that replaces `SpanAndGuard` that was used in conjunction with `rental`. 

This allows us to remove the now unmaintained `rental` crate.

### CODE REVIEW

File: primitives/tracing/src/proxied_span.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+//! Comprises the struct `ProxiedTrace` required by the proxy, encapsulates
+//! the necessary associated unsafe code and provides a safe interface.
+
+use std::pin::Pin;
+use tracing::{span::Entered, Span};
+
+/// Container for a proxied tracing Span and it's associated guard
+pub struct ProxiedSpan {
+	id: u64,
+	// Note that `guard` and `span` must be declared in this order so that `guard` is dropped first.
```
Comment (bkchr) [inline primitives/tracing/src/proxied_span.rs:26]: Please write a custom `Drop` impl so ensure this!

Comment (mattrutherford) [reply primitives/tracing/src/proxied_span.rs]: Struct field drop order is now stable <https://doc.rust-lang.org/reference/destructors.html>, <https://github.com/rust-lang/rfcs/pull/1857> but agree it's good to be explicit about this. In terms of explicitness it seems like `ManuallyDrop` is a better choice than `Option::take()`, even though it requires another block of unsafe code.

Comment (bkchr) [reply primitives/tracing/src/proxied_span.rs]: Yeah I know that this is defined, but better safe than sorry :D

**Comment by mattrutherford** [comment]
This new code essentially replicates what the rental crate was doing, just with a strictly limited interface. I think it should soon be possible to replace the need for this altogether, so perhaps it's worth leaving things as is with the rental crate for now. However I don't think there would be any downside to merging this, if deemed desirable.

File: primitives/tracing/src/proxied_span.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+impl ProxiedSpan {
+	/// Enter the supplied span and return a new instance of ProxiedTrace containing it
+	pub fn enter_span(id: u64, span: Pin<Box<Span>>) -> Self {
+		let span = ManuallyDrop::new(span);
+		// Transmute to static lifetime to enable guard to be stored
+		// along with the span that it references
+		let guard = unsafe {
+			ManuallyDrop::new(transmute::<Entered<'_>, Entered<'static>>(span.enter()))
```
Comment (kianenigma) [inline primitives/tracing/src/proxied_span.rs:39]: can you explain a bit more what is going on here?

**Comment by bkchr** [comment]
@mattrutherford Do we still need this?

**Comment by mattrutherford** [comment]
It's uncertain, but possible that we still need to do this for WASM, because as of right now, experiments to access the spans directly via a new interface breaks the relationship between spans from wasm and corresponding native tracing spans/events. It's possible to avoid this using our customer subscriber if we switch to use 2 Events (enter/exit) instead of a single Span and patch them to our span representation in the subscriber, but that wouldn't preserve the parent_id relationships if we adopt the tracing_subscriber::FmtSubscriber

**Pull request closed without merging**